lib/TypeNames/DependencyGraph.cpp

```diff
@@ -57,216 +63,470 @@ std::string llvm::DOTGraphTraits<DepGraph *>::getNodeLabel(const DepNode *N,
+    addAndLogSuccessor(DependentDecl, Specifier.Declaration);
+
+  // If we found a pointer on the path to TypeSpecifierDef, the definition of
+  // Dependent only needs the declaration of the Specifier.
+  // Otherwise it needs the full definition of the Specifier.
+  if (FoundPointer)
+    addAndLogSuccessor(DependentDef, Specifier.Declaration);
+  else
+    addAndLogSuccessor(DependentDef, Specifier.Definition);
+}
+
+DependencyGraph::AssociatedNodes
+DGBuilder::addWrappedRFWithDependencies(const model::RawFunctionDefinition &RF)
+  const {
+  AssociatedNodes ReturnValuesWrapper = addArtificialNodes(RF);
+  const auto &[RVWrapperDecl, RVWrapperDef] = ReturnValuesWrapper;
+
+  using abi::FunctionType::Layout;
+  auto Layout = Layout::make(RF);
+
+  // Then the artificial wrapper wrapping the return values should depend on the
+  // relevant return values types, that are the types of its fields.
+  auto ReturnValues = returnValueTypes(Layout);
+  revng_assert(not ReturnValues.empty());
+  for (const model::UpcastableType &ReturnType : ReturnValues) {
+    revng_assert(not ReturnType->isArray());
+    addDependenciesFrom(ReturnValuesWrapper, *ReturnType);
+  }
+
+  return ReturnValuesWrapper;
+}
+
+void DGBuilder::addDependencies(const model::TypeDefinition &T) const {
 
-      if (auto *D = getDependencyFor<TypeNode::Declaration>(*Edge,
-                                                            TypeToNode)) {
-        Deps.push_back({ DeclNode, D });
-        revng_log(Log,
-                  getNodeLabel(DeclNode) << " depends on " << getNodeLabel(D));
+  const auto &TDNodes = Graph->TypeToNodes.at(&T);
+  const auto &[DeclNode, DefNode] = TDNodes;
+
+  switch (T.Kind()) {
+
+  case model::TypeDefinitionKind::TypedefDefinition:
+  case model::TypeDefinitionKind::EnumDefinition:
+  case model::TypeDefinitionKind::StructDefinition:
+  case model::TypeDefinitionKind::UnionDefinition: {
+    for (const model::Type *Edge : T.edges())
+      addDependenciesFrom(TDNodes, *Edge);
+  } break;
+
+  case model::TypeDefinitionKind::RawFunctionDefinition:
+  case model::TypeDefinitionKind::CABIFunctionDefinition: {
+    // Function types may require to add struct wrappers around argument types
+    // and return types.
+    using abi::FunctionType::Layout;
+    auto Layout = Layout::make(T);
+
+    using namespace abi::FunctionType::ReturnMethod;
+    if (Layout.returnMethod() == RegisterSet) {
+
+      // If T is a RawFunctionDefinition returning a RegisterSet, we create an
+      // artificial struct wrapper around the returned registers.
+      const auto *RF = cast<model::RawFunctionDefinition>(&T);
+      AssociatedNodes WrapperNodes = addWrappedRFWithDependencies(*RF);
+
+      // Both the declaration and the definition of the function type depend
+      // only on the declaration of the struct wrapper for the return type.
+      addAndLogSuccessor(DeclNode, WrapperNodes.Declaration);
+      addAndLogSuccessor(DefNode, WrapperNodes.Declaration);
```

Comment:
Wait, is there even a point of adding `DefNode -> WrapperNodes.Declaration` edge? It's already there transitively through `DefNode -> DeclNode -> WrapperNodes.Declaration`.